segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MultiKeyReaderWriterScheduler.java

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.metadata;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A scheduler utility that implements pattern similar to Multiple Readers - Single Writer pattern.
+ */
+public class MultiKeyReaderWriterScheduler {
+    /**
+     * Scheduler data for each key.
+     */
+    private final HashMap<String, SchedulerData> keyToDataMap = new HashMap<>();
```

Comment:
Please use `@GuardedBy` on the fields that need synchronization on, then in your IDE settings, make sure you enable `Concurrency warnings`. The IDE will tell you every single invocation where you accessing this outside of declared guards.

Reply:
done